core/services/pipeline/orm.go

```diff
@@ -23,7 +23,7 @@ type (
 	ORM interface {
 		CreateSpec(ctx context.Context, db *gorm.DB, taskDAG TaskDAG, maxTaskTimeout models.Interval) (int32, error)
 		CreateRun(ctx context.Context, jobID int32, meta map[string]interface{}) (int64, error)
-		ProcessNextUnclaimedTaskRun(ctx context.Context, fn ProcessTaskRunFunc) (bool, error)
+		ProcessNextUnfinishedRun(ctx context.Context, fn ProcessRunFunc) (bool, error)
```

Comment:
I think we can even go further to simplify this - if we're fine just having jobs executed by a single node, I think the whole save->trigger notify->pick up notification->read job, execute and save flow isn't needed anymore. 

Observe and similar scenarios would just use the ExecuteTaskRuns directly like (psuedo code)

```
func (ds dataSource) Observe(ctx context.Context) (ocrtypes.Observation, error) {
        1. Open a transaction
        2. Read the job/pipeline spec/task specs with ds.jobID and lock that job (actually may not even need the lock since only one node is running the libocr.Oracle at a time?). Something like pipelineORM.GetPipeline(tx, jobID) 
        3. ExecuteTaskRuns (in memory, returns a pipeline.Run with its TaskRuns populated and completed)
        4. Insert the pipeline.Run and its associated task runs. Something like pipelineORM.SaveRun(tx, Run)
        5. Close transaction
}
```
Totally something that can be done as a follow up. However I think it might save you some testing re-working if we do that simplification up front

Reply:
Lets talk about this on slack

Reply:
Lets talk about this on slack

Reply:
Implementing this in a followup PR